File: rb/ArtInstituteOfChicago_sdk.rb
# ArtInstituteOfChicago SDK

require_relative 'utility/struct/voxgig_struct'
require_relative 'core/utility_type'
require_relative 'core/spec'
require_relative 'core/helpers'

# Load utility registration
require_relative 'utility/register'

# Load config and features
require_relative 'config'
require_relative 'feature/base_feature'
require_relative 'features'

# Load typed models (Struct value objects).
require_relative 'ArtInstituteOfChicago_types'


class ArtInstituteOfChicagoSDK
  attr_accessor :mode, :features, :options

  def initialize(options = {})
    @mode = "live"
    @features = []
    @options = nil

    utility = ArtInstituteOfChicagoUtility.new
    @_utility = utility

    config = ArtInstituteOfChicagoConfig.make_config

    @_rootctx = utility.make_context.call({
      "client" => self,
      "utility" => utility,
      "config" => config,
      "options" => options || {},
      "shared" => {},
    }, nil)

    @options = utility.make_options.call(@_rootctx)

    if VoxgigStruct.getpath(@options, "feature.test.active") == true
      @mode = "test"
    end

    @_rootctx.options = @options

    # Add features from config.
    feature_opts = ArtInstituteOfChicagoHelpers.to_map(VoxgigStruct.getprop(@options, "feature"))
    if feature_opts
      items = VoxgigStruct.items(feature_opts)
      if items
        items.each do |item|
          fname = item[0]
          fopts = ArtInstituteOfChicagoHelpers.to_map(item[1])
          if fopts && fopts["active"] == true
            utility.feature_add.call(@_rootctx, ArtInstituteOfChicagoFeatures.make_feature(fname))
          end
        end
      end
    end

    # Add extension features.
    extend_val = VoxgigStruct.getprop(@options, "extend")
    if extend_val.is_a?(Array)
      extend_val.each do |f|
        if f.respond_to?(:get_name)
          utility.feature_add.call(@_rootctx, f)
        end
      end
    end

    # Initialize features.
    @features.each do |f|
      utility.feature_init.call(@_rootctx, f)
    end

    utility.feature_hook.call(@_rootctx, "PostConstruct")
  end

  def options_map
    out = VoxgigStruct.clone(@options)
    out.is_a?(Hash) ? out : {}
  end

  def get_utility
    ArtInstituteOfChicagoUtility.copy(@_utility)
  end

  def get_root_ctx
    @_rootctx
  end

  def prepare(fetchargs = {})
    utility = @_utility
    fetchargs ||= {}

    ctrl = ArtInstituteOfChicagoHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "prepare",
      "ctrl" => ctrl,
    }, @_rootctx)

    opts = @options
    path = VoxgigStruct.getprop(fetchargs, "path") || ""
    path = "" unless path.is_a?(String)
    method_val = VoxgigStruct.getprop(fetchargs, "method") || "GET"
    method_val = "GET" unless method_val.is_a?(String)
    params = ArtInstituteOfChicagoHelpers.to_map(VoxgigStruct.getprop(fetchargs, "params")) || {}
    query = ArtInstituteOfChicagoHelpers.to_map(VoxgigStruct.getprop(fetchargs, "query")) || {}
    headers = utility.prepare_headers.call(ctx)

    base = VoxgigStruct.getprop(opts, "base") || ""
    base = "" unless base.is_a?(String)
    prefix = VoxgigStruct.getprop(opts, "prefix") || ""
    prefix = "" unless prefix.is_a?(String)
    suffix = VoxgigStruct.getprop(opts, "suffix") || ""
    suffix = "" unless suffix.is_a?(String)

    ctx.spec = ArtInstituteOfChicagoSpec.new({
      "base" => base, "prefix" => prefix, "suffix" => suffix,
      "path" => path, "method" => method_val,
      "params" => params, "query" => query, "headers" => headers,
      "body" => VoxgigStruct.getprop(fetchargs, "body"),
      "step" => "start",
    })

    # Merge user-provided headers.
    uh = VoxgigStruct.getprop(fetchargs, "headers")
    if uh.is_a?(Hash)
      uh.each { |k, v| ctx.spec.headers[k] = v }
    end

    _, err = utility.prepare_auth.call(ctx)
    raise err if err

    # make_fetch_def returns a (fetchdef, err) tuple; destructure it and
    # return just the fetchdef Hash (raising on error) so callers — including
    # direct(), which indexes fetchdef["url"] — receive a Hash, mirroring the
    # ts/py prepare().
    fetchdef, fd_err = utility.make_fetch_def.call(ctx)
    raise fd_err if fd_err

    fetchdef
  end

  def direct(fetchargs = {})
    utility = @_utility

    # direct() is the raw-HTTP escape hatch: it always returns a result hash
    # ({ "ok" => ..., ... }) and never raises. prepare() raises on error, so
    # trap that and surface it in the hash.
    begin
      fetchdef = prepare(fetchargs)
    rescue ArtInstituteOfChicagoError => err
      return { "ok" => false, "err" => err }
    end

    fetchargs ||= {}
    ctrl = ArtInstituteOfChicagoHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "direct",
      "ctrl" => ctrl,
    }, @_rootctx)

    url = fetchdef["url"] || ""
    fetched, fetch_err = utility.fetcher.call(ctx, url, fetchdef)

    return { "ok" => false, "err" => fetch_err } if fetch_err

    if fetched.nil?
      return {
        "ok" => false,
        "err" => ctx.make_error("direct_no_response", "response: undefined"),
      }
    end

    if fetched.is_a?(Hash)
      status = ArtInstituteOfChicagoHelpers.to_int(VoxgigStruct.getprop(fetched, "status"))
      headers = VoxgigStruct.getprop(fetched, "headers") || {}

      # No-body responses (204, 304) and explicit zero content-length must
      # skip JSON parsing — calling json() on an empty body errors.
      content_length = headers.is_a?(Hash) ? headers["content-length"] : nil
      no_body = status == 204 || status == 304 || content_length.to_s == "0"

      json_data = nil
      unless no_body
        jf = VoxgigStruct.getprop(fetched, "json")
        if jf.is_a?(Proc)
          begin
            json_data = jf.call
          rescue StandardError
            # Non-JSON body — leave data nil, keep status/headers.
            json_data = nil
          end
        end
      end

      return {
        "ok" => status >= 200 && status < 300,
        "status" => status,
        "headers" => headers,
        "data" => json_data,
      }
    end

    return {
      "ok" => false,
      "err" => ctx.make_error("direct_invalid", "invalid response type"),
    }
  end


  # Canonical facade: client.Agent.list / client.Agent.load({ "id" => ... })
  def Agent(data = nil)
    require_relative 'entity/agent_entity'
    AgentEntity.new(self, data)
  end


  # Canonical facade: client.AgentRole.list / client.AgentRole.load({ "id" => ... })
  def AgentRole(data = nil)
    require_relative 'entity/agent_role_entity'
    AgentRoleEntity.new(self, data)
  end


  # Canonical facade: client.AgentType.list / client.AgentType.load({ "id" => ... })
  def AgentType(data = nil)
    require_relative 'entity/agent_type_entity'
    AgentTypeEntity.new(self, data)
  end


  # Canonical facade: client.Article.list / client.Article.load({ "id" => ... })
  def Article(data = nil)
    require_relative 'entity/article_entity'
    ArticleEntity.new(self, data)
  end


  # Canonical facade: client.Artwork.list / client.Artwork.load({ "id" => ... })
  def Artwork(data = nil)
    require_relative 'entity/artwork_entity'
    ArtworkEntity.new(self, data)
  end


  # Canonical facade: client.ArtworkDateQualifier.list / client.ArtworkDateQualifier.load({ "id" => ... })
  def ArtworkDateQualifier(data = nil)
    require_relative 'entity/artwork_date_qualifier_entity'
    ArtworkDateQualifierEntity.new(self, data)
  end


  # Canonical facade: client.ArtworkPlaceQualifier.list / client.ArtworkPlaceQualifier.load({ "id" => ... })
  def ArtworkPlaceQualifier(data = nil)
    require_relative 'entity/artwork_place_qualifier_entity'
    ArtworkPlaceQualifierEntity.new(self, data)
  end


  # Canonical facade: client.ArtworkType.list / client.ArtworkType.load({ "id" => ... })
  def ArtworkType(data = nil)
    require_relative 'entity/artwork_type_entity'
    ArtworkTypeEntity.new(self, data)
  end


  # Canonical facade: client.CategoryTerm.list / client.CategoryTerm.load({ "id" => ... })
  def CategoryTerm(data = nil)
    require_relative 'entity/category_term_entity'
    CategoryTermEntity.new(self, data)
  end


  # Canonical facade: client.DigitalPublication.list / client.DigitalPublication.load({ "id" => ... })
  def DigitalPublication(data = nil)
    require_relative 'entity/digital_publication_entity'
    DigitalPublicationEntity.new(self, data)
  end


  # Canonical facade: client.DigitalPublicationArticle.list / client.DigitalPublicationArticle.load({ "id" => ... })
  def DigitalPublicationArticle(data = nil)
    require_relative 'entity/digital_publication_article_entity'
    DigitalPublicationArticleEntity.new(self, data)
  end


  # Canonical facade: client.EducatorResource.list / client.EducatorResource.load({ "id" => ... })
  def EducatorResource(data = nil)
    require_relative 'entity/educator_resource_entity'
    EducatorResourceEntity.new(self, data)
  end


  # Canonical facade: client.Event.list / client.Event.load({ "id" => ... })
  def Event(data = nil)
    require_relative 'entity/event_entity'
    EventEntity.new(self, data)
  end


  # Canonical facade: client.EventOccurrence.list / client.EventOccurrence.load({ "id" => ... })
  def EventOccurrence(data = nil)
    require_relative 'entity/event_occurrence_entity'
    EventOccurrenceEntity.new(self, data)
  end


  # Canonical facade: client.EventProgram.list / client.EventProgram.load({ "id" => ... })
  def EventProgram(data = nil)
    require_relative 'entity/event_program_entity'
    EventProgramEntity.new(self, data)
  end


  # Canonical facade: client.Exhibition.list / client.Exhibition.load({ "id" => ... })
  def Exhibition(data = nil)
    require_relative 'entity/exhibition_entity'
    ExhibitionEntity.new(self, data)
  end


  # Canonical facade: client.Gallery.list / client.Gallery.load({ "id" => ... })
  def Gallery(data = nil)
    require_relative 'entity/gallery_entity'
    GalleryEntity.new(self, data)
  end


  # Canonical facade: client.GenericPage.list / client.GenericPage.load({ "id" => ... })
  def GenericPage(data = nil)
    require_relative 'entity/generic_page_entity'
    GenericPageEntity.new(self, data)
  end


  # Canonical facade: client.Highlight.list / client.Highlight.load({ "id" => ... })
  def Highlight(data = nil)
    require_relative 'entity/highlight_entity'
    HighlightEntity.new(self, data)
  end


  # Canonical facade: client.Hour.list / client.Hour.load({ "id" => ... })
  def Hour(data = nil)
    require_relative 'entity/hour_entity'
    HourEntity.new(self, data)
  end


  # Canonical facade: client.Image.list / client.Image.load({ "id" => ... })
  def Image(data = nil)
    require_relative 'entity/image_entity'
    ImageEntity.new(self, data)
  end


  # Canonical facade: client.LandingPage.list / client.LandingPage.load({ "id" => ... })
  def LandingPage(data = nil)
    require_relative 'entity/landing_page_entity'
    LandingPageEntity.new(self, data)
  end


  # Canonical facade: client.Place.list / client.Place.load({ "id" => ... })
  def Place(data = nil)
    require_relative 'entity/place_entity'
    PlaceEntity.new(self, data)
  end


  # Canonical facade: client.PressRelease.list / client.PressRelease.load({ "id" => ... })
  def PressRelease(data = nil)
    require_relative 'entity/press_release_entity'
    PressReleaseEntity.new(self, data)
  end


  # Canonical facade: client.PrintedPublication.list / client.PrintedPublication.load({ "id" => ... })
  def PrintedPublication(data = nil)
    require_relative 'entity/printed_publication_entity'
    PrintedPublicationEntity.new(self, data)
  end


  # Canonical facade: client.Product.list / client.Product.load({ "id" => ... })
  def Product(data = nil)
    require_relative 'entity/product_entity'
    ProductEntity.new(self, data)
  end


  # Canonical facade: client.Publication.list / client.Publication.load({ "id" => ... })
  def Publication(data = nil)
    require_relative 'entity/publication_entity'
    PublicationEntity.new(self, data)
  end


  # Canonical facade: client.Search.list / client.Search.load({ "id" => ... })
  def Search(data = nil)
    require_relative 'entity/search_entity'
    SearchEntity.new(self, data)
  end


  # Canonical facade: client.Section.list / client.Section.load({ "id" => ... })
  def Section(data = nil)
    require_relative 'entity/section_entity'
    SectionEntity.new(self, data)
  end


  # Canonical facade: client.Site.list / client.Site.load({ "id" => ... })
  def Site(data = nil)
    require_relative 'entity/site_entity'
    SiteEntity.new(self, data)
  end


  # Canonical facade: client.Sound.list / client.Sound.load({ "id" => ... })
  def Sound(data = nil)
    require_relative 'entity/sound_entity'
    SoundEntity.new(self, data)
  end


  # Canonical facade: client.StaticPage.list / client.StaticPage.load({ "id" => ... })
  def StaticPage(data = nil)
    require_relative 'entity/static_page_entity'
    StaticPageEntity.new(self, data)
  end


  # Canonical facade: client.Text.list / client.Text.load({ "id" => ... })
  def Text(data = nil)
    require_relative 'entity/text_entity'
    TextEntity.new(self, data)
  end


  # Canonical facade: client.Tour.list / client.Tour.load({ "id" => ... })
  def Tour(data = nil)
    require_relative 'entity/tour_entity'
    TourEntity.new(self, data)
  end


  # Canonical facade: client.Video.list / client.Video.load({ "id" => ... })
  def Video(data = nil)
    require_relative 'entity/video_entity'
    VideoEntity.new(self, data)
  end



  def self.test(testopts = nil, sdkopts = nil)
    sdkopts = sdkopts || {}
    sdkopts = VoxgigStruct.clone(sdkopts)
    sdkopts = {} unless sdkopts.is_a?(Hash)

    testopts = testopts || {}
    testopts = VoxgigStruct.clone(testopts)
    testopts = {} unless testopts.is_a?(Hash)
    testopts["active"] = true

    VoxgigStruct.setpath(sdkopts, "feature.test", testopts)

    sdk = ArtInstituteOfChicagoSDK.new(sdkopts)
    sdk.mode = "test"
    sdk
  end
end

File: py/artinstituteofchicago_sdk.py
# ArtInstituteOfChicago SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import ArtInstituteOfChicagoUtility
from core.spec import ArtInstituteOfChicagoSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import ArtInstituteOfChicagoBaseFeature
from features import _make_feature


class ArtInstituteOfChicagoSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = ArtInstituteOfChicagoUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return ArtInstituteOfChicagoUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = ArtInstituteOfChicagoSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    def Agent(self, data=None) -> "AgentEntity":
        """Entity factory: client.Agent().list() / client.Agent().load({"id": ...})."""
        from entity.agent_entity import AgentEntity
        return AgentEntity(self, data)


    def AgentRole(self, data=None) -> "AgentRoleEntity":
        """Entity factory: client.AgentRole().list() / client.AgentRole().load({"id": ...})."""
        from entity.agent_role_entity import AgentRoleEntity
        return AgentRoleEntity(self, data)


    def AgentType(self, data=None) -> "AgentTypeEntity":
        """Entity factory: client.AgentType().list() / client.AgentType().load({"id": ...})."""
        from entity.agent_type_entity import AgentTypeEntity
        return AgentTypeEntity(self, data)


    def Article(self, data=None) -> "ArticleEntity":
        """Entity factory: client.Article().list() / client.Article().load({"id": ...})."""
        from entity.article_entity import ArticleEntity
        return ArticleEntity(self, data)


    def Artwork(self, data=None) -> "ArtworkEntity":
        """Entity factory: client.Artwork().list() / client.Artwork().load({"id": ...})."""
        from entity.artwork_entity import ArtworkEntity
        return ArtworkEntity(self, data)


    def ArtworkDateQualifier(self, data=None) -> "ArtworkDateQualifierEntity":
        """Entity factory: client.ArtworkDateQualifier().list() / client.ArtworkDateQualifier().load({"id": ...})."""
        from entity.artwork_date_qualifier_entity import ArtworkDateQualifierEntity
        return ArtworkDateQualifierEntity(self, data)


    def ArtworkPlaceQualifier(self, data=None) -> "ArtworkPlaceQualifierEntity":
        """Entity factory: client.ArtworkPlaceQualifier().list() / client.ArtworkPlaceQualifier().load({"id": ...})."""
        from entity.artwork_place_qualifier_entity import ArtworkPlaceQualifierEntity
        return ArtworkPlaceQualifierEntity(self, data)


    def ArtworkType(self, data=None) -> "ArtworkTypeEntity":
        """Entity factory: client.ArtworkType().list() / client.ArtworkType().load({"id": ...})."""
        from entity.artwork_type_entity import ArtworkTypeEntity
        return ArtworkTypeEntity(self, data)


    def CategoryTerm(self, data=None) -> "CategoryTermEntity":
        """Entity factory: client.CategoryTerm().list() / client.CategoryTerm().load({"id": ...})."""
        from entity.category_term_entity import CategoryTermEntity
        return CategoryTermEntity(self, data)


    def DigitalPublication(self, data=None) -> "DigitalPublicationEntity":
        """Entity factory: client.DigitalPublication().list() / client.DigitalPublication().load({"id": ...})."""
        from entity.digital_publication_entity import DigitalPublicationEntity
        return DigitalPublicationEntity(self, data)


    def DigitalPublicationArticle(self, data=None) -> "DigitalPublicationArticleEntity":
        """Entity factory: client.DigitalPublicationArticle().list() / client.DigitalPublicationArticle().load({"id": ...})."""
        from entity.digital_publication_article_entity import DigitalPublicationArticleEntity
        return DigitalPublicationArticleEntity(self, data)


    def EducatorResource(self, data=None) -> "EducatorResourceEntity":
        """Entity factory: client.EducatorResource().list() / client.EducatorResource().load({"id": ...})."""
        from entity.educator_resource_entity import EducatorResourceEntity
        return EducatorResourceEntity(self, data)


    def Event(self, data=None) -> "EventEntity":
        """Entity factory: client.Event().list() / client.Event().load({"id": ...})."""
        from entity.event_entity import EventEntity
        return EventEntity(self, data)


    def EventOccurrence(self, data=None) -> "EventOccurrenceEntity":
        """Entity factory: client.EventOccurrence().list() / client.EventOccurrence().load({"id": ...})."""
        from entity.event_occurrence_entity import EventOccurrenceEntity
        return EventOccurrenceEntity(self, data)


    def EventProgram(self, data=None) -> "EventProgramEntity":
        """Entity factory: client.EventProgram().list() / client.EventProgram().load({"id": ...})."""
        from entity.event_program_entity import EventProgramEntity
        return EventProgramEntity(self, data)


    def Exhibition(self, data=None) -> "ExhibitionEntity":
        """Entity factory: client.Exhibition().list() / client.Exhibition().load({"id": ...})."""
        from entity.exhibition_entity import ExhibitionEntity
        return ExhibitionEntity(self, data)


    def Gallery(self, data=None) -> "GalleryEntity":
        """Entity factory: client.Gallery().list() / client.Gallery().load({"id": ...})."""
        from entity.gallery_entity import GalleryEntity
        return GalleryEntity(self, data)


    def GenericPage(self, data=None) -> "GenericPageEntity":
        """Entity factory: client.GenericPage().list() / client.GenericPage().load({"id": ...})."""
        from entity.generic_page_entity import GenericPageEntity
        return GenericPageEntity(self, data)


    def Highlight(self, data=None) -> "HighlightEntity":
        """Entity factory: client.Highlight().list() / client.Highlight().load({"id": ...})."""
        from entity.highlight_entity import HighlightEntity
        return HighlightEntity(self, data)


    def Hour(self, data=None) -> "HourEntity":
        """Entity factory: client.Hour().list() / client.Hour().load({"id": ...})."""
        from entity.hour_entity import HourEntity
        return HourEntity(self, data)


    def Image(self, data=None) -> "ImageEntity":
        """Entity factory: client.Image().list() / client.Image().load({"id": ...})."""
        from entity.image_entity import ImageEntity
        return ImageEntity(self, data)


    def LandingPage(self, data=None) -> "LandingPageEntity":
        """Entity factory: client.LandingPage().list() / client.LandingPage().load({"id": ...})."""
        from entity.landing_page_entity import LandingPageEntity
        return LandingPageEntity(self, data)


    def Place(self, data=None) -> "PlaceEntity":
        """Entity factory: client.Place().list() / client.Place().load({"id": ...})."""
        from entity.place_entity import PlaceEntity
        return PlaceEntity(self, data)


    def PressRelease(self, data=None) -> "PressReleaseEntity":
        """Entity factory: client.PressRelease().list() / client.PressRelease().load({"id": ...})."""
        from entity.press_release_entity import PressReleaseEntity
        return PressReleaseEntity(self, data)


    def PrintedPublication(self, data=None) -> "PrintedPublicationEntity":
        """Entity factory: client.PrintedPublication().list() / client.PrintedPublication().load({"id": ...})."""
        from entity.printed_publication_entity import PrintedPublicationEntity
        return PrintedPublicationEntity(self, data)


    def Product(self, data=None) -> "ProductEntity":
        """Entity factory: client.Product().list() / client.Product().load({"id": ...})."""
        from entity.product_entity import ProductEntity
        return ProductEntity(self, data)


    def Publication(self, data=None) -> "PublicationEntity":
        """Entity factory: client.Publication().list() / client.Publication().load({"id": ...})."""
        from entity.publication_entity import PublicationEntity
        return PublicationEntity(self, data)


    def Search(self, data=None) -> "SearchEntity":
        """Entity factory: client.Search().list() / client.Search().load({"id": ...})."""
        from entity.search_entity import SearchEntity
        return SearchEntity(self, data)


    def Section(self, data=None) -> "SectionEntity":
        """Entity factory: client.Section().list() / client.Section().load({"id": ...})."""
        from entity.section_entity import SectionEntity
        return SectionEntity(self, data)


    def Site(self, data=None) -> "SiteEntity":
        """Entity factory: client.Site().list() / client.Site().load({"id": ...})."""
        from entity.site_entity import SiteEntity
        return SiteEntity(self, data)


    def Sound(self, data=None) -> "SoundEntity":
        """Entity factory: client.Sound().list() / client.Sound().load({"id": ...})."""
        from entity.sound_entity import SoundEntity
        return SoundEntity(self, data)


    def StaticPage(self, data=None) -> "StaticPageEntity":
        """Entity factory: client.StaticPage().list() / client.StaticPage().load({"id": ...})."""
        from entity.static_page_entity import StaticPageEntity
        return StaticPageEntity(self, data)


    def Text(self, data=None) -> "TextEntity":
        """Entity factory: client.Text().list() / client.Text().load({"id": ...})."""
        from entity.text_entity import TextEntity
        return TextEntity(self, data)


    def Tour(self, data=None) -> "TourEntity":
        """Entity factory: client.Tour().list() / client.Tour().load({"id": ...})."""
        from entity.tour_entity import TourEntity
        return TourEntity(self, data)


    def Video(self, data=None) -> "VideoEntity":
        """Entity factory: client.Video().list() / client.Video().load({"id": ...})."""
        from entity.video_entity import VideoEntity
        return VideoEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None) -> "ArtInstituteOfChicagoSDK":
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk


from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from entity.agent_entity import AgentEntity
    from entity.agent_role_entity import AgentRoleEntity
    from entity.agent_type_entity import AgentTypeEntity
    from entity.article_entity import ArticleEntity
    from entity.artwork_entity import ArtworkEntity
    from entity.artwork_date_qualifier_entity import ArtworkDateQualifierEntity
    from entity.artwork_place_qualifier_entity import ArtworkPlaceQualifierEntity
    from entity.artwork_type_entity import ArtworkTypeEntity
    from entity.category_term_entity import CategoryTermEntity
    from entity.digital_publication_entity import DigitalPublicationEntity
    from entity.digital_publication_article_entity import DigitalPublicationArticleEntity
    from entity.educator_resource_entity import EducatorResourceEntity
    from entity.event_entity import EventEntity
    from entity.event_occurrence_entity import EventOccurrenceEntity
    from entity.event_program_entity import EventProgramEntity
    from entity.exhibition_entity import ExhibitionEntity
    from entity.gallery_entity import GalleryEntity
    from entity.generic_page_entity import GenericPageEntity
    from entity.highlight_entity import HighlightEntity
    from entity.hour_entity import HourEntity
    from entity.image_entity import ImageEntity
    from entity.landing_page_entity import LandingPageEntity
    from entity.place_entity import PlaceEntity
    from entity.press_release_entity import PressReleaseEntity
    from entity.printed_publication_entity import PrintedPublicationEntity
    from entity.product_entity import ProductEntity
    from entity.publication_entity import PublicationEntity
    from entity.search_entity import SearchEntity
    from entity.section_entity import SectionEntity
    from entity.site_entity import SiteEntity
    from entity.sound_entity import SoundEntity
    from entity.static_page_entity import StaticPageEntity
    from entity.text_entity import TextEntity
    from entity.tour_entity import TourEntity
    from entity.video_entity import VideoEntity
